### PR TITLE
fix a bug about snapshot and adjust some test cases

### PR DESCRIPTION
**Pull request by hicqu**
The bug causes an exmaple panics in #205 .

In our implemetation, `RaftLog::term(un_received_index)` returns `0`. So if a peer receives a snapshot or normal entry with log-term `0`, `RaftLog::match_term(log_term, un_received_index)` will return `true` but it should be `false` exactly.

This PR fixes it by making `RaftLog::match_term` cover this case.

And, add a case `test_snapshot_with_term_0` to cover it, and do some adjusts for some other test cases. At last, remove `Interface::initial` because we don't need it any more.

### CODE REVIEW

**Comment by hicqu** [comment]
PTAL @Hoverbear @BusyJay thank you!

File: examples/five_mem_node/main.rs [hunk 1]
```diff
@@ -207,6 +211,15 @@ fn on_ready(
         return;
     }
 
+    // Apply the snashot. It's also necessary with same reason as above.
```
Comment (Hoverbear) [inline examples/five_mem_node/main.rs]:
```suggestion
    // Apply the snapshot. It's also necessary with same reason as above.
```

Also, which reason? That the program will be blocked forever?

File: harness/src/network.rs [hunk 2]
```diff
@@ -61,21 +61,39 @@ pub struct Network {
 }
 
 impl Network {
-    /// Initializes a network from peers.
+    /// Get a base config. Calling `Network::new` will initialize peers with this config.
+    pub fn base_config() -> Config {
```
Comment (Hoverbear) [inline harness/src/network.rs]:
```suggestion
    pub fn default_test_config() -> Config {
```

**Comment by Hoverbear** [comment]
@hicqu Seems maybe `RaftLog::term(un_received_index)` should return `None` instead? Would that be a good way to fix this problem?

**Comment by hicqu** [comment]
Thanks for your review @Hoverbear ! I will address them in #214 . Now I will close this PR because I prefer the another solution. :)

**Comment by bytewatch** [comment]
I think cause the root cause of the panic of five_mem_node is MemStorage::initialize_with_conf_state() :

```
        // To fix this panic,  add a more line: `core.raft_state.hard_state.set_commit(1)` , 
        // and handle the case that `prs.get_mut(self.id)` is None in restore_snapshot.
        // Or , just commet these two lines: 
        core.snapshot_metadata.set_index(1);  
        core.raft_state.hard_state.set_commit(1);  
        core.raft_state.conf_state = ConfState::from(conf_state);

```

File: src/raft_log.rs [hunk 3]
```diff
@@ -199,7 +199,12 @@ impl<T: Storage> RaftLog<T> {
 
     /// Answers the question: Does this index belong to this term?
     pub fn match_term(&self, idx: u64, term: u64) -> bool {
-        self.term(idx).map(|t| t == term).unwrap_or(false)
+        match self.term(idx) {
+            // For uninitialized storage, should return false.
+            Ok(0) => term == 0 && self.last_index() > 0,
```
Comment (bytewatch) [inline src/raft_log.rs]: I don't think it need to add some logics that exceeds etcd-raft.

**Pull request closed without merging**